community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/StoreCopyCheckPointMutex.java

```diff
@@ -0,0 +1,199 @@
+ *            |                                   |
+ *            <-CHECKPOINT--|-STORE-COPY---------->
+ * </pre>
+ *
+ * In the image above there are three "events":
+ * <ol>
+ * <li>Store-copy 1, where there are three concurrent store-copies going on.
+ * Only the first one performs check-point</li>
+ * <li>External check-point, which waits for the ongoing store-copies to complete and then performs it</li>
+ * <li>Store-copy 2, which waits for the external check-point to complete and then starts its own
+ * check-point, which is part of the store-copy algorithm to then perform the store-copy.</li>
+ * </ol>
+ *
+ * Status changes are made in synchronized as opposed to atomic CAS operations, since this results
+ * in simpler code and since this mutex is normally called a couple of times per hour it's not an issue.
+ */
+public class StoreCopyCheckPointMutex
+{
+    /**
+     * Main lock. Read-lock is for {@link #storeCopy(ThrowingAction)} and write-lock is for {@link #checkPoint()}.
+     */
+    private final ReadWriteLock lock;
+
+    /**
+     * Number of currently ongoing store-copy requests.
+     */
+    private int storeCopyCount;
+
+    /**
+     * Whether or not the first (of the concurrently ongoing store-copy requests) has had its "before"
+     * action completed. The other store-copy requests will wait for this flag to be {@code true}.
+     */
+    private volatile boolean storeCopyActionCompleted;
+
+    /**
+     * Error which may have happened during first concurrent store-copy request. Made available to
+     * the other concurrent store-copy requests so that they can fail instead of waiting forever.
+     */
+    private volatile Throwable storeCopyActionError;
+
+    public StoreCopyCheckPointMutex()
+    {
+        this( new ReentrantReadWriteLock( true ) );
+    }
+
+    public StoreCopyCheckPointMutex( ReadWriteLock lock )
+    {
+        this.lock = lock;
+    }
+
+    public Resource storeCopy( ThrowingAction<IOException> beforeFirstConcurrentStoreCopy ) throws IOException
+    {
+        Lock readLock = lock.readLock();
+        boolean firstConcurrentRead = incrementCount() == 0;
+        boolean success = false;
+        try
+        {
+            if ( firstConcurrentRead )
+            {
+                try
+                {
+                    beforeFirstConcurrentStoreCopy.apply();
+                }
+                catch ( Throwable e )
+                {
+                    storeCopyActionError = e;
+                    throw launderedException( IOException.class, e );
+                }
+                storeCopyActionCompleted = true;
+            }
+            else
+            {
+                // Wait for the "before" first store copy to complete
+                waitForFirstStoreCopyActionToComplete();
+            }
+            success = true;
+        }
+        finally
+        {
+            if ( success )
+            {
+                readLock.lock();
```

Comment:
Yes, and I think that's fine because either the store-copy check-point will get first and the external check-point will have to wait for store-copy session to finish, or the external check-point will win and the store-copy check-point will come after it. In both cases the second check-point should be quite fast. I don't think functionality this is a problem.